vllm/v1/spec_decode/ngram_proposer_gpu.py

```diff
@@ -0,0 +1,465 @@
+        assert vllm_config.speculative_config.prompt_lookup_max is not None
+
+        self.min_n = vllm_config.speculative_config.prompt_lookup_min
+        self.max_n = vllm_config.speculative_config.prompt_lookup_max
+        self.k = vllm_config.speculative_config.num_speculative_tokens
+        self.max_model_len = vllm_config.model_config.max_model_len
+        self.max_num_seqs = vllm_config.scheduler_config.max_num_seqs
+        self.vocab_size = vllm_config.model_config.get_vocab_size()
+        self.device = device
+
+    def _find_first_and_extract_all_n_parallel(
+        self,
+        data: torch.Tensor,
+        seq_lengths: torch.Tensor,
+        min_pattern_len: int,
+        max_pattern_len: int,
+        result_len: int,
+    ) -> torch.Tensor:
+        """
+        Process all pattern lengths in parallel, selecting the longest match.
+        Completely free of data-dependent control flow, suitable for
+        torch.compile optimization.
+
+        This function allows partial results when there are fewer than result_len
+        tokens available after the match. The remaining positions are filled with 0.
+        """
+        batch_size = data.shape[0]
+        device = data.device
+        max_seq_len = data.shape[1]
+        num_patterns = max_pattern_len - min_pattern_len + 1
+
+        all_windows = data.unfold(1, max_pattern_len, 1)  # [B, num_windows, max_n]
+        num_windows = all_windows.shape[1]
+        window_starts = torch.arange(num_windows, device=device)
+        pattern_lengths = torch.arange(
+            min_pattern_len, max_pattern_len + 1, device=device
+        )
+        batch_indices = torch.arange(batch_size, device=device)
+
+        all_first_matches = torch.full(
+            (batch_size, num_patterns), -1, dtype=torch.long, device=device
+        )
+
+        for i, pattern_len in enumerate(range(min_pattern_len, max_pattern_len + 1)):
+            offset = max_pattern_len - pattern_len
+
+            # Extract pattern from the end of each sequence
+            pattern_starts = seq_lengths - pattern_len
+            pattern_indices = pattern_starts.unsqueeze(1) + torch.arange(
+                pattern_len, device=device
+            )
+            patterns = torch.gather(data, 1, pattern_indices.clamp(min=0))
+
+            # Slice windows and perform matching
+            current_windows = all_windows[..., offset:]
+            matches = (current_windows == patterns.unsqueeze(1)).all(dim=-1)
+
+            # Validity check: ensure enough space for result extraction
+            max_valid_pattern_start = seq_lengths - pattern_len - 1
+            pattern_start_positions = window_starts + offset
+            valid_mask = pattern_start_positions <= max_valid_pattern_start.unsqueeze(1)
+            final_matches = matches & valid_mask
+
+            # Handle prefix positions that fall before the available windows
+            if offset > 0:
```

Comment:
If I understand correctly, this code handles looking for matches in the first `offset` tokens of the prompt, since `all_windows[..., offset:]` is derived from `data.unfold(1, max_pattern_len, 1)` which expands the input tokens into windows of size `max_pattern_len` and therefore slicing can miss the first few patterns of size `pattern_len`.

I think it is much simpler and more efficient to just have `current_windows = data.unfold(1, pattern_len, 1)` inside the loop, which will automatically cover that edge-case that you currently are handling manually.